libcudacxx/codegen/generators/compare_and_swap.h

```diff
@@ -0,0 +1,175 @@
+        case __ATOMIC_RELEASE: __res = __cuda_cas(__atomic_cuda_release{}); break;
+        case __ATOMIC_RELAXED: __res = __cuda_cas(__atomic_cuda_relaxed{}); break;
+        default: assert(0);
+      }
+    ),
+    NV_IS_DEVICE, (
+      switch (__stronger_order_cuda(__success_memorder, __failure_memorder)) {
+        case __ATOMIC_SEQ_CST: _CCCL_FALLTHROUGH();
+        case __ATOMIC_ACQ_REL: __cuda_atomic_membar(_Sco{}); _CCCL_FALLTHROUGH();
+        case __ATOMIC_CONSUME: _CCCL_FALLTHROUGH();
+        case __ATOMIC_ACQUIRE: __res = __cuda_cas(__atomic_cuda_volatile{}); __cuda_atomic_membar(_Sco{}); break;
+        case __ATOMIC_RELEASE: __cuda_atomic_membar(_Sco{}); __res = __cuda_cas(__atomic_cuda_volatile{}); break;
+        case __ATOMIC_RELAXED: __res = __cuda_cas(__atomic_cuda_volatile{}); break;
+        default: assert(0);
+      }
+    )
+  )
+  return __res;
+}
+)XXX";
+
+  // Argument ID Reference
+  // 0 - Operand Type
+  // 1 - Operand Size
+  // 2 - Type Constraint
+  // 3 - Memory Order
+  // 4 - Memory Order function tag
+  // 5 - Scope Constraint
+  // 6 - Scope function tag
+  const std::string asm_intrinsic_format_128 = R"XXX(
+template <class _Type>
+static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
+  _Type* __ptr, _Type& __dst, _Type __cmp, _Type __op, {4}, __atomic_cuda_operand_{0}{1}, {6})
+{{
+  asm volatile(R"YYY(
+.reg .b128 _d;
+.reg .b128 _v;
+mov.b128 {{%0, %1}}, _d;
+mov.b128 {{%4, %5}}, _v;
+atom.cas{3}{5}.b128 _d,[%2],_d,_v;
+mov.b128 _d, {{%0, %1}};
+)YYY" : "=l"(__dst.__x),"=l"(__dst.__y) : "l"(__ptr), "l"(__cmp.__x),"l"(__cmp.__y), "l"(__op.__x),"l"(__op.__y) : "memory"); return __dst.x == __cmp.x && __dst.y == __cmp.y; }})XXX";
+
+  const std::string asm_intrinsic_format = R"XXX(
+template <class _Type>
+static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
+  _Type* __ptr, _Type& __dst, _Type __cmp, _Type __op, {4}, __atomic_cuda_operand_{0}{1}, {6})
+{{ asm volatile("atom.cas{3}{5}.{0}{1} %0,[%1],%2,%3;" : "={2}"(__dst) : "l"(__ptr), "{2}"(__cmp), "{2}"(__op) : "memory"); return __dst == __cmp; }})XXX";
+
+  constexpr Operand supported_types[] = {
+    Operand::Bit,
+  };
+
+  constexpr size_t supported_sizes[] = {
+    16,
+    32,
+    64,
+    128,
+  };
+
+  constexpr Semantic supported_semantics[] = {
+    Semantic::Acquire,
+    Semantic::Relaxed,
+    Semantic::Release,
+    Semantic::Acq_Rel,
+    Semantic::Volatile,
+  };
+
+  constexpr Scope supported_scopes[] = {
+    Scope::CTA,
+    Scope::Cluster,
+    Scope::GPU,
+    Scope::System,
+  };
+
+  for (auto size : supported_sizes)
+  {
+    for (auto type : supported_types)
+    {
+      for (auto sem : supported_semantics)
+      {
+        for (auto sco : supported_scopes)
+        {
+          if (size == 2 && type != Operand::Bit)
+          {
+            continue;
+          }
+          if (size == 128 && type != Operand::Bit)
+          {
+            continue;
+          }
+          out << fmt::format(
```

Comment:
Suggestion: print to the ostream directly
```suggestion
          fmt::print(out, 
```

Reply:
Can't print directly to the `ostream` from `fmt::` on MSVC. I assume that it's just broken as it keeps trying to use the string API as though there isn't an overload defined for it.

Reply:
Sad, but thanks for trying!